### PR TITLE
aii-opennebula: Add an option to change user/group resources

### LINKED ISSUE #173: aii-opennebula: Add an option to assign the new cloud resources to a specific user/group
At this moment all the resources managed and generated by the AII are assigned to oneadmin user. Add an option to choose instead oneadmin a different user or group.

### PR DESCRIPTION
**Pull request by alvarosimon**
* Fix #173 : Add an option to assign the new cloud resources to a specific user/group

**NOTE:** This PR requires Net::OpenNebula 0.306.0 to use  `.chmod` and `.chown` methods.

### CODE REVIEW

File: aii-opennebula/src/main/pan/quattor/aii/opennebula/schema.pan [hunk 1]
```diff
@@ -107,10 +107,21 @@ type opennebula_ignoremac = {
     "interface" ? valid_interface_ignoremac[]
 };
 
+@documentation{
+Type that changes resources owner/group permissions.
+By default opennebula-aii generates all the resources as oneadmin owner/group.
+}
+type opennebula_permissions = {
+    "user"  ? string # id or user name
```
Comment (stdweird) [inline aii-opennebula/src/main/pan/quattor/aii/opennebula/schema.pan]: use `owner` if you can.

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 2]
```diff
@@ -13,7 +13,7 @@ use Set::Scalar;
 use Template;
 
 use Config::Tiny;
-use Net::OpenNebula 0.2.2;
+use Net::OpenNebula 0.306.0;
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: use `0.307.0` which has the speed fix for the poolinfo queries

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 3]
```diff
@@ -180,16 +192,75 @@ sub new
     return bless {}, $class;
 }
 
+sub change_permissions
+{
+    my ($self, $one, $type, $resource, $permissions) = @_;
+    my ($method, $id, $instance, %chown, $out);
+    my @options = qw(user group mode);
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: owner instead of user

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 4]
```diff
@@ -180,16 +192,75 @@ sub new
...
+        if (exists($permissions->{$value})) {
+            if ($value eq "mode") {
+                $out = $resource->chmod($permissions->{$value});
+                if ($out) {
+                    $main::this_app->info("Changed $type mode id $out to: ", $permissions->{$value});
+                } else {
+                    $main::this_app->error("Not able to change $type mode to: ", $permissions->{$value});
+                };
+            } else {
+                if ($value eq "user") {
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: owner

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 5]
```diff
@@ -180,16 +192,75 @@ sub new
...
+    my ($method, $id, $instance, %chown, $out);
+    my @options = qw(user group mode);
+    # -1 value by default, owner and group are not changed
+    %chown = (
+        uid => -1,
+        gid => -1,
+        one => $one,
+    );
+
+    foreach my $value (@options) {
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: this looks very complex with the for loop. why not
```perl
my %chown = (one => $one);
my $mode = $permissions->{mode};
if(defined($mode)) {
    $out = $resource->chmod($mode);
    if ($out) {
        $main::this_app->info("Changed $type mode id $out to: $mode");
    } else {
        $main::this_app->error("Not able to change $type mode to: $mode");
    }
}
$chown{uid} = defined($permissions->{owner}) $permissions->{owner} : -1;
$chown{gid} = defined($permissions->{group}) $permissions->{group} : -1;
```

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 6]
```diff
@@ -180,16 +192,75 @@ sub new
...
+                    $chown{uid} = $permissions->{$value};
+                } else {
+                    $chown{gid} = $permissions->{$value};
+                };
+            };
+        };
+    };
+
+    $out = $resource->chown(%chown);
+    if ($out) {
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: use a `my $msg = "user:group $chown{uid}:$chown{gid} for: " . $resource->name;` for th ecommon message part

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 7]
```diff
@@ -180,16 +192,75 @@ sub new
...
+
+    $method = "get_users" if ($resource eq "owner");
+
+    my @existres = $one->$method(qr{^$name$});
+
+    foreach my $t (@existres) {
+        $main::this_app->info("Found requested $resource in ONE database: $name");
+        return $t;
+    };
+    $main::this_app->error("Not able to found $resource name $name in ONE database");
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: `found` -> `find`

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 8]
```diff
@@ -220,6 +291,9 @@ sub create_vm_images
         }
         if ($newimage) {
             $main::this_app->info("Created new VM image ID: ", $newimage->id);
+            if (ref $permissions eq ref {}) {
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: is same as `if ($permissions)`

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 9]
```diff
@@ -415,6 +489,9 @@ sub remove_and_create_vm_template
                 $main::this_app->info("QUATTOR VM template, going to update: ",$t->name);
                 $self->debug(1, "New $fqdn template : $vmtemplate");
                 my $update = $t->update($vmtemplate, 0);
+                if (ref $permissions eq ref {}) {
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: `if ($permissions)`

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 10]
```diff
@@ -567,6 +649,10 @@ sub install
         my $vmid = $vmtemplate->instantiate(name => $fqdn, onhold => $onhold);
         if (defined($vmid) && $vmid =~ m/^\d+$/) {
             $main::this_app->info("VM ${fqdn} was created successfully with ID: ${vmid}");
+            if (ref $permissions eq ref {}) {
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: `if ($permissions)`

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 11]
```diff
@@ -619,7 +705,7 @@ sub remove
 
     my %images = $self->get_images($config);
     if (%images) {
-        $self->remove_and_create_vm_images($one, 1, \%images, $rmimage);
+        $self->remove_and_create_vm_images($one, 1, \%images, 0, $rmimage);
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: pass `undef` instead of `0`

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 12]
```diff
@@ -629,7 +715,7 @@ sub remove
 
     my $vmtemplatetxt = $self->get_vmtemplate($config);
     if ($vmtemplatetxt) {
-        $self->remove_and_create_vm_template($one, $fqdn, 1, $vmtemplatetxt, $rmvmtemplate);
+        $self->remove_and_create_vm_template($one, $fqdn, 1, $vmtemplatetxt, 0, $rmvmtemplate);
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: `undef` instead of `0`

**Comment by stdweird** [comment]
retest this please

File: aii-opennebula/src/main/perl/opennebula.pm (hunk 5 above)
Comment (alvarosimon) [inline aii-opennebula/src/main/perl/opennebula.pm]: @stdweird much better indeed :smile:

File: aii-opennebula/src/main/perl/opennebula.pm (hunk 8 above)
Comment (alvarosimon) [inline aii-opennebula/src/main/perl/opennebula.pm]: @stdweird It's strange in that case I get this error:
`Can't use string ("1") as a HASH ref while "strict refs" in use at /home/lsimngar/UGent/hpcugent/aii/aii-opennebula/target/lib/perl/AII/opennebula.pm`

from:
`$chown{uid} = defined($permissions->{owner}) ? $permissions->{owner} : -1;`

That is the reason why I have included this condition to be sure that I'm getting a HASH, maybe something is wrong in the code, I have to check....

**Comment by alvarosimon** [comment]
@stdweird included remarks and tested in altaria

File: aii-opennebula/src/main/pan/quattor/aii/opennebula/schema.pan [hunk 13]
```diff
@@ -107,10 +107,21 @@ type opennebula_ignoremac = {
     "interface" ? valid_interface_ignoremac[]
 };
 
+@documentation{
+Type that changes resources owner/group permissions.
+By default opennebula-aii generates all the resources as oneadmin owner/group.
+}
+type opennebula_permissions = {
+    "owner"  ? string # id or user name
```
Comment (stdweird) [inline aii-opennebula/src/main/pan/quattor/aii/opennebula/schema.pan]: add the comment in an annotation `@{}`, and clarify the the `id` here is the (internal) ONE id of the user.

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 14]
```diff
@@ -91,6 +91,18 @@ sub process_template
     return "$tpl";
 }
 
+sub get_permissions
+{
+    my ($self, $config) = @_;
+
+    my $tree = $config->getElement('/system/opennebula')->getTree();
+    if ($tree->{permissions}) {
+        $main::this_app->info("Found new resources permissions.");
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: maybe add the permissions to the message?

File: aii-opennebula/src/main/perl/opennebula.pm [hunk 15]
```diff
@@ -180,16 +192,61 @@ sub new
...
+    my ($method, $id, $instance, $out);
+    my %chown = (one => $one);
+    my $mode = $permissions->{mode};
+
+    if(defined($mode)) {
+        $out = $resource->chmod($mode);
+        if ($out) {
+            $main::this_app->info("Changed $type mode id $out to: $mode");
+        } else {
+        $main::this_app->error("Not able to change $type mode to: $mode");
```
Comment (stdweird) [inline aii-opennebula/src/main/perl/opennebula.pm]: fix indentation

**Comment by stdweird** [comment]
@jrha @jouvin i bumped this to 16.6, as it fixes the failing aii tests too

**Comment by stdweird** [comment]
LGTM

**Pull request merged**